### PR TITLE
[Blog Post] Instrumenting Phoenix with Telemetry Part I

### PR DESCRIPTION
**Pull request by SophieDeBenedetto**
First of a series of four blog posts on this topic, as discussed in this issue https://github.com/elixirschool/elixirschool/issues/2244

I tried to strike a balance between "how to use Telemetry" and "how it works under the hood" but not sure if I lost the narrative somewhere in there--all feedback super welcome as always!

~Currently working on drafts of the next three posts, all together the series will cover:~
UPDATE: This PR includes drafts for the first post, and the other  three posts are PR'd in the links below

* Part I: Telemetry Under The Hood
* [Part II: Handling Telemetry Events with `TelemetryMetrics` + `TelemetryMetricsStatsd`](https://github.com/elixirschool/elixirschool/pull/2284)
* [Part III: Observing Phoenix + Ecto Telemetry Events](https://github.com/elixirschool/elixirschool/pull/2285)
* [Part IV: Erlang VM Measurements with `telemetry_poller`, `TelemetryMetrics` + `TelemetryMetricsStatsd`](https://github.com/elixirschool/elixirschool/pull/2286)

### CODE REVIEW

File: _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+
+In Part I we'll start out by setting up a basic, DIY Telemetry pipeline and examining how Erlang's Telemetry library works under the hood. Then, in Part II we'll take advantage of the `TelemetryMetrics` and `TelemetryMetricsStatsd` libraries to respond to Telemetry events by formatting them as metrics and reporting those metrics to StatsD. In Part III, we'll look at the powerful instrumentation that Phoenix and Ecto offer out-of-the-box via Telemetry events executed in source code. Lastly, in Part IV, we'll leverage the `telemetry_poller` Erlang library to take Erlang VM measurements and emit them as Telemetry events, which our Telemetry pipeline can then observe and report on.
+
+## Introduction
+
+In this post we'll discuss why observability matters and how Telemetry helps us treat observability like a first class citizen in Elixir projects. Then, we'll hand-roll our own instrumentation pipeline using Telemetry and StatsD. We'll wrap up with a look under the hood of the Telemetry library and set ourselves for Part II of this series, in which we leverage the `Telemetry.Metrics` library for even easier instrumentation and reporting.
+
+## Observability Matters
```
Comment (SophieDeBenedetto) [inline _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md]: Maybe too much exposition? I could cut down this section a fair amount I think.

File: _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+
+Now that we understand why observability matters and how Telemetry supports observability, let's dive into the Telemetry library!
+
+## Using Telemetry in Elixir
+
+First, we'll take a look at how to set up a simple reporting pipeline for custom Telemetry events in our Phoenix app. Then, we'll take a look under the hood at the Telemetry library to understand how our pipeline works.
+
+### Getting Started
+
+You can follow along with this tutorial by cloning down the repo [here](https://github.com/SophieDeBenedetto/quantum/tree/part-1-start).
```
Comment (SophieDeBenedetto) [inline _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md]: If/when we publish I can move this example app repo to `elixirschool` org.

**Comment by jordelver** [comment]
I don't have any feedback other than to say thank you for writing this, I look forward to reading it once published :)

File: _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+* Checkig out the starting state of our code on the branch [part-1-start](https://github.com/SophieDeBenedetto/quantum/tree/part-1-start)
```
Comment (vbrazo) [inline _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md]:
```suggestion
* Checking out the starting state of our code on the branch [part-1-start](https://github.com/SophieDeBenedetto/quantum/tree/part-1-start)
```

**Comment by SophieDeBenedetto** [comment]
I added all four posts in one PR b/c I though it might be easier to review all in one place but I know its A LOT! Do folks think we should:
* Publish the series all together, take the extra time to review things up-front
* Publish posts one at a time, over a few weeks

If the second option, I can split up the articles into four separate PRs to make it easier to review and publish.

File: _posts/instrumenting_phoenix_with_telemetry_part_three.md [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+
+We can add some metrics reporting for these events by defining metrics in `Quantum.Telemetry` for either of the `"phoenix.channel_joined"` and `"phoenix.channel_handled_in"` events.
+
+Now that we've taken a brief tour of Phoenix Telemetry events, let's hook up some reporting for Ecto events.
+
+## Ecto Telemetry Events
+
+Ecto provides some out-of-the-box instrumentation for queries. Let's take a look at and define metrics for some of these Telemetry events now.
+
+Ecto will execute a Telemetry event, [`[:my_app, :repo, :query]`](https://github.com/elixir-ecto/ecto/blob/2aca7b28eef486188be66592055c7336a80befe9/lib/ecto/repo.ex#L117) for every query sent to the Ecto adapter. It will emit this event with a measurement map and a metadata map.
```
Comment (SophieDeBenedetto) [inline _posts/instrumenting_phoenix_with_telemetry_part_three.md:382]: Change this to link to published docs on Hex, not docs in source code.

**Comment by SophieDeBenedetto** [comment]
Thanks for your suggestions @tajchumber! Ready for a final 👀

File: _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md (hunk 1 above)
Comment (SophieDeBenedetto) [reply _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md:27]: I cut a bit

File: _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md (hunk 2 above)
Comment (SophieDeBenedetto) [reply _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md:59]: This is done and the links are updated

File: _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md (hunk 3 above)
Comment (SophieDeBenedetto) [reply _posts/2020-03-29-instrumenting-phoenix-with-telemetry-part-one.md]: fixed!

**Pull request merged**